profiles/audit_log.py

```diff
@@ -37,56 +86,109 @@ def _profile_part(instance):
+        and _should_audit(instance.__class__)
         and instance.pk
     ):
-        logger = logging.getLogger("audit")
-
-        current_time = datetime.now(tz=timezone.utc)
-        current_user = get_current_user()
         profile = (
             instance.profile
             if hasattr(instance, "profile")
             else instance.resolve_profile()
         )
-        profile_id = str(profile.pk) if profile else None
-        target_user = profile.user if profile and profile.user else None
-
-        message = {
-            "audit_event": {
-                "origin": "PROFILE-BE",
-                "status": "SUCCESS",
-                "date_time_epoch": int(current_time.timestamp() * 1000),
-                "date_time": f"{current_time.replace(tzinfo=None).isoformat(sep='T', timespec='milliseconds')}Z",
-                "actor": {"role": _resolve_role(current_user, profile)},
-                "operation": action,
-                "target": {"id": profile_id, "type": _profile_part(instance)},
-            }
+
+        profile_loggables = audit_loggables[profile.pk]
+
+        if action == "DELETE" and isinstance(instance, Profile):
+            # If the Profile is about to get deleted,
+            # need to store the User before it gets deleted too.
+            profile_loggables["user"] = profile.user
+        profile_loggables["profile"] = profile
+        profile_loggables["parts"].add((action, _profile_part(instance)))
+
+
+def _produce_json_logs(
+    current_user, current_time, service, client_id, ip_address, audit_loggables
+):
+    logger = logging.getLogger("audit")
+
+    date_time_epoch = int(current_time.timestamp() * 1000)
+    date_time = f"{current_time.replace(tzinfo=None).isoformat(sep='T', timespec='milliseconds')}Z"
+
+    actor_obj = {}
+
+    if service:
+        actor_obj["service_name"] = service.name
+
+    if client_id:
+        actor_obj["client_id"] = client_id
+
+    if ip_address:
+        actor_obj["ip_address"] = ip_address
+
+    _format_user_data(actor_obj, current_user)
+
+    for profile_id, data in audit_loggables.items():
+        target_user = data.get("user")
+
+        actor_obj["role"] = _resolve_role(current_user, target_user)
+
+        target_obj = {
+            "id": str(profile_id),
         }
+        _format_user_data(target_obj, target_user)
+
+        for action, profile_part in data["parts"]:
+            target_obj["type"] = profile_part
+
+            message = {
+                "audit_event": {
+                    "origin": "PROFILE-BE",
+                    "status": "SUCCESS",
+                    "date_time_epoch": date_time_epoch,
+                    "date_time": date_time,
+                    "actor": actor_obj,
+                    "operation": action,
+                    "target": target_obj,
+                }
+            }
 
-        _format_user_data(message["audit_event"], "actor", current_user)
+            logger.info(json.dumps(message))
```

Comment:
At least currently all the to-be-encoded data is supplied by the audit logging system itself and there's no user supplied data. So I don't know how the JSON encoder could fail, unless it's running out of memory or some such general error case.

But if we think about the upcoming feature that audit log write error should be such a fatal error that it even prevents the user getting any data (or in case of database writes the transactions should be rolled back, which feature we don't yet off course have), then we probably shouldn't at least silently swallow any error happening in the audit logging system, right?